File: 1466-jump-game-v/jump-game-v.cpp
class Solution {
public:
    int maxJumps(vector<int>& arr, int d) {
        int n = arr.size();
        vector<int> dp(n,-1);

        function<int(int)>dfs = [&](int i){
            if (dp[i]!=-1)
                return dp[i];
            int best=1;

            for(int j=i+1; j<=min(n-1,i+d);++j){
                if (arr[j]>=arr[i])
                    break;
                best = max(best,1 + dfs(j));
            }

            for (int j=i-1; j>=max(0,i-d);--j){
                if (arr[j]>=arr[i])
                    break;
                best = max(best,1 + dfs(j));
            }
            dp[i]=best;
            return dp[i];
        };

        int ans=0;
        for(int i=0;i<n;++i)
            ans = max(ans,dfs(i));
        return ans;
    }
};

File: 1466-jump-game-v/jump-game-v.py
class Solution:
    def maxJumps(self, arr: List[int], d: int) -> int:
        n = len(arr)
        dp = [-1]*n

        def dfs(i):
            if dp[i] != -1:
                return dp[i]

            best=1

            for j in range(i+1, min(n,i+d+1)):
                if (arr[j]>=arr[i]):
                    break
                best = max(best,1+dfs(j))

            for j in range(i-1, max(-1,i-d-1), -1):
                if (arr[j]>=arr[i]):
                    break
                best = max(best,1+dfs(j))

            dp[i] = best
            return best

        ans=0
        for i in range(n):
            ans = max(ans,dfs(i))
        return ans
        